feat(stock): algorithm for quantity in stock

jniles opened the pull request:
Adds an algorithm developed by @jeremielodi and formatted by @jniles for computing the quantity in stock of inventory items.

Closes #4864.

jniles committed:
feat(stock): algorithm for quantity in stock

Adds an algorithm developed by @jeremielodi and formatted by @jniles for
computing the quantity in stock of inventory items.

Closes #4864.

jniles committed:
feat(stock): algorithm for quantity in stock

Adds an algorithm developed by @jeremielodi and formatted by @jniles for
computing the quantity in stock of inventory items.

Closes #4864.

jmcameron commented on server/models/admin.sql:
```diff
@@ -194,4 +194,36 @@ CREATE PROCEDURE zMergeAccounts(
+DROP PROCEDURE IF EXISTS zComputeAllInventoryStockQuantities $$
+CREATE PROCEDURE zComputeAllInventoryStockQuantities (
+  IN _start_date DATE,
+  IN _depot_uuid BINARY(16)
```
Aren't we going to need an end date too?   For instance, we may need to compute these things over a month but may be several days into the next month.

jmcameron commented on server/models/procedures/stock.sql:
```diff
@@ -274,4 +274,188 @@ BEGIN
+  DECLARE _end_date DATE;
+  DECLARE _qtt, _in_qtt, _out_qtt DECIMAL(19, 4);
+
+  -- DROP TEMPORARY TABLE IF EXISTS `stock_movement_status`;
```
Don't you want to drop this table in the main routine above?  Or is "temporary table" enough to avoid reusing the table between queries?

jeremielodi replied:
It's should be a permanent table, I used a temporary table just for the demostration

jeremielodi replied:
@jniles we don't need to store the ```inventory name``` as we already have it ```uuid```

jmcameron replied:
Thanks for the explanations.  That helped me understand what you were doing here.

jmcameron commented on server/models/procedures/stock.sql:
```diff
@@ -274,4 +274,188 @@ BEGIN
+  END IF;
+
+  SET @depot_counter = 1;
+  -- let's loop throw all inventories
```
throw -> through

jmcameron reviewed (approved):
I was impressed by this SQL code.  Nice job @jeremielodi !

I'm pretty sure we will want to be able to specify the end date for these searches.   Someone wanting to do them through the current date can just pass in now().

Could you give some statistics about how long this takes to run on the various databases (please include the number of  moments involved so we get a sense of how this scales).    

I found the labels for the aggregate data a little confusing.  A possible suggestion for making that clearer

- in_quantity => start_quantity
- out_quantity => consumed  (or quantity_consumed)
- in_stock_quantity => end_quantity

(I presume that is what these columns mean.)
Note that these names correspond with the "Start Date" and "End Date"

jeremielodi reviewed (commented):
Good work @jniles, I wonder why you don't call the procedure ```computeStockQuantity``` after a stock movement registration.

mbayopanda reviewed (approved):
The procedures work without errors, I am wondering just we have there temporary tables so how can we correctly test the procedure ?

lomamech commented on server/models/procedures/stock.sql:
```diff
@@ -274,4 +274,188 @@ BEGIN
+*/
+DROP PROCEDURE IF EXISTS `computeStockQuantity`$$
+CREATE PROCEDURE `computeStockQuantity` (
+  IN _start_date DATE,
```
If there will not be an end date, the best will be to simply use date instead of start date

lomamech commented on server/models/procedures/stock.sql:
```diff
@@ -274,4 +274,188 @@ BEGIN
+  DECLARE _qtt, _in_qtt, _out_qtt DECIMAL(19, 4);
+
+  -- DROP TEMPORARY TABLE IF EXISTS `stock_movement_status`;
+  CREATE TEMPORARY TABLE IF NOT EXISTS `stock_movement_status` (
```
but also since we will use a permanent table instead of a temporary table I find that it is a good proposal

jeremielodi committed:
calling cmm data calculating for new stock movement

jeremielodi committed:
calling cmm data calculating for new stock movement

jniles committed:
perf(stock): use inventory for quantity calcs

This commit removes the lotUuids and prevents duplicate calls to the
computeStockQuantity() by filtering out duplicate inventoryUuids.

jeremielodi committed:
get CMM procedure

jniles committed:
Merge remote-tracking branch 'jeremie/cmm-data-population' into 4864-feat-compute-quanity-in-stock-proceedure

jniles committed:
fix(adjustment): add inventory to data.

Ships back the inventory_uuid for convenience during the creation of an
adjustment writing.  Also adds documenting comments.

jniles committed:
fix(schema): FK constraints stock_movement_status

Adds foreign key constraints to `stock_movement_status`.

mbayopanda commented on server/controllers/stock/index.js:
```diff
@@ -52,85 +52,118 @@ exports.getStockTransfers = getStockTransfers;
+
+  // makes a unique array of inventory uuids so we don't do extra calls
+  const uniqueInventoryUuids = inventoryUuids
+    .filter((uid, index, array) => array.lastIndexOf(uid) === index);
```
🐐

jeremielodi replied:
I know why this error occur(in_quantity=0 and out_quantity =0), I will try to solve the issue when it will be created

jeremielodi replied:
As this PR is not merged yet @jniles can update the procedure ```ComputeStockQuantiy``` as follow 

```sql


DROP PROCEDURE IF EXISTS `computeStockQuantity`$$
CREATE PROCEDURE `computeStockQuantity` (
  IN _start_date DATE,
  IN _inventory_uuid BINARY(16),
  IN _depot_filter_uuid BINARY(16)
) BEGIN
  DECLARE done BOOLEAN;
  DECLARE _depot_uuid, _row_uuid BINARY(16);
  DECLARE _end_date DATE;
  DECLARE _qtt, _in_qtt, _out_qtt DECIMAL(19, 4);




  DROP TEMPORARY TABLE IF EXISTS `temp_stock_movement`;

    -- every stock movement for a inventory will be stored here in order to facilitate search
  CREATE TEMPORARY TABLE `temp_stock_movement` (
    `reference` INT,
    `date` DATE,
    `depot_uuid` BINARY(16),
    `inventory_uuid` BINARY(16),
    `quantity` DECIMAL(19,4),
    `in_quantity` DECIMAL(19,4),
    `out_quantity` DECIMAL(19,4),
    KEY `depot_uuid` (`depot_uuid`),
    KEY `inventory_uuid` (`inventory_uuid`),
    KEY `reference` (`reference`)
  ) ENGINE=InnoDB DEFAULT CHARACTER SET = utf8mb4 DEFAULT COLLATE = utf8mb4_unicode_ci;

  -- every stock movement for a inventory will be stored here in order to facilitate search
  CREATE TEMPORARY TABLE IF NOT EXISTS `temp_depot` (
    `reference` INT,
    `uuid` BINARY(16),
    `text` varchar(100),
    KEY `uuid` (`uuid`),
    KEY `reference` (`reference`)
  ) ENGINE=InnoDB DEFAULT CHARACTER SET = utf8mb4 DEFAULT COLLATE = utf8mb4_unicode_ci;

  SET _end_date = NOW();
  SET _start_date = DATE(_start_date);
  SET @depot_number = 0;

  DELETE FROM `temp_depot`;

  SET @filter_by_depot = (_depot_filter_uuid IS NOT NULL);

  IF @filter_by_depot = 0  THEN -- will compute inventory's quantity in each depot
    INSERT INTO `temp_depot`
    SELECT (@depot_number:=@depot_number + 1)as ref, `uuid`, `text`
    FROM depot;
  ELSE
    INSERT INTO `temp_depot`  -- will compute inventory's quantity just in a specific depot
    SELECT (@depot_number := @depot_number + 1) as ref, `uuid`, `text`
    FROM depot
    WHERE `uuid` = _depot_filter_uuid;
  END IF;

  SET @depot_counter = 1;
  -- let's loop through all inventories

  WHILE (@depot_counter <= @depot_number) DO
    SET @depot_uuid = (SELECT `uuid` FROM `temp_depot` WHERE `reference` = @depot_counter);

    -- delete all inventory's data for this period

    DELETE FROM `stock_movement_status`
    WHERE  `start_date` >= _start_date AND  `end_date` <= _end_date AND  `inventory_uuid` = _inventory_uuid AND `depot_uuid` =  @depot_uuid;

    DELETE FROM `temp_stock_movement` WHERE 1;
    INSERT INTO `temp_stock_movement`

    SELECT (@row_number:=@row_number + 1) AS `reference`, x.* FROM (
      SELECT DATE(m.date) AS `date`,  m.depot_uuid, i.uuid,
        SUM( IFNULL(m.quantity * IF(m.is_exit = 1, -1, 1), 0)) AS quantity,
        SUM(IF(m.is_exit = 0, m.quantity, 0)),
        SUM(IF(m.is_exit = 1, m.quantity, 0))
      FROM `stock_movement` m
        JOIN lot l ON l.uuid = m.lot_uuid
        JOIN inventory i ON i.uuid = l.inventory_uuid
      WHERE i.uuid = _inventory_uuid AND DATE(m.date) <= DATE(_end_date) AND m.depot_uuid =  @depot_uuid
      GROUP BY  DATE(m.date), m.depot_uuid
      ORDER BY  `date`
    ) AS x
    ORDER BY x.date;

	 
    SET @row_i = 0;
    SET _row_uuid = NULL;

    SELECT SUM(quantity) AS quantity INTO _qtt
    FROM temp_stock_movement m
      JOIN inventory i ON i.uuid = m.inventory_uuid
      JOIN depot d ON d.uuid = m.depot_uuid
    WHERE DATE(m.date) <=_start_date AND m.depot_uuid = @depot_uuid;

    SET _qtt = IFNULL(_qtt, 0);

      SELECT in_quantity, out_quantity 
      INTO _in_qtt, _out_qtt
      FROM temp_stock_movement m
      WHERE DATE(m.date) =DATE(_start_date) AND m.depot_uuid = @depot_uuid; 
 
    DELETE FROM temp_stock_movement WHERE DATE(`date`) <=_start_date AND `depot_uuid` = @depot_uuid;

    -- check if this date already exist in stock_movement_status for the inventory
    SET @date_exists = 0;

    SELECT `uuid`, count(`uuid`) as nbr
      INTO _row_uuid,  @date_exists
    FROM `stock_movement_status`
    WHERE `depot_uuid` = @depot_uuid AND `inventory_uuid` = _inventory_uuid
      AND DATE(`start_date`) <= _start_date
    ORDER BY `start_date` DESC
    LIMIT 1;

    IF @date_exists = 0 THEN
      SET _row_uuid  = HUID(uuid());
      INSERT INTO  `stock_movement_status` VALUES (_row_uuid, _start_date, _start_date, _qtt, IFNULL(_in_qtt, 0), IFNULL(_out_qtt, 0), _inventory_uuid, @depot_uuid);
    END IF;

    SELECT `reference` INTO @row_i
    FROM temp_stock_movement
    WHERE `date` > _start_date
    LIMIT 1;

    SET @mvts_number = (SELECT COUNT(*) from temp_stock_movement  WHERE `date` > _start_date);
    SET  @row_number = @row_i + @mvts_number - 1;

    SET @row_i = IFNULL(@row_i, 0);

    IF @row_i = 0 THEN
      UPDATE `stock_movement_status` SET `end_date` = _end_date WHERE `uuid` = _row_uuid;
    ELSE
      WHILE (@row_i < @row_number + 1) DO
        SET _in_qtt = 0;
        SET _out_qtt = 0;

        SET @current_qtt = 0;
        SET @current_date = _start_date;

        SELECT  m.quantity, m.in_quantity, m.out_quantity, m.date
          INTO @current_qtt, _in_qtt, _out_qtt, @current_date
        FROM temp_stock_movement m
        JOIN inventory i ON i.uuid = m.inventory_uuid
        JOIN depot d ON d.uuid = m.depot_uuid
        WHERE m.reference = @row_i
        GROUP BY m.depot_uuid;
        UPDATE `stock_movement_status` SET `end_date` = DATE_SUB(@current_date, INTERVAL 1 DAY) WHERE `uuid` = _row_uuid;
        IF @current_qtt <> _qtt THEN
          SET _row_uuid = HUID(UUID());
          SET _qtt = @current_qtt + _qtt;
          INSERT INTO  `stock_movement_status` VALUES (_row_uuid, @current_date, @current_date, _qtt, _in_qtt, _out_qtt, _inventory_uuid,  @depot_uuid);
        END IF;
        SET @row_i= @row_i + 1;
      END WHILE;
    END IF;

    UPDATE `stock_movement_status` SET `end_date` = _end_date WHERE `uuid` = _row_uuid;
    SET _qtt = 0;
    SET @depot_counter = @depot_counter +1;
    SET _row_uuid = NULL;
  END WHILE;
END$$

```

Before the line where we delete some movement in ```temp_stock_movement ``` that we won't need anymore
I get the in and out quantity of the start_date 
```sql     

      
      SELECT in_quantity, out_quantity 
      INTO _in_qtt, _out_qtt
      FROM temp_stock_movement m
      WHERE DATE(m.date) =DATE(_start_date) AND m.depot_uuid = @depot_uuid; 
 
    DELETE FROM temp_stock_movement WHERE DATE(`date`) <=_start_date AND `depot_uuid` = @depot_uuid;
```
After that I check in the ```stock_movement_status``` if there is a quantity computation of this inventory at that date(start_date), In the issue of @mbayopanda nothing will be found because the record of ```2020-09-01``` was deleted in order to recompute.
You should know that the procedure delete every row in ```stock_movement_status``` for the given period

To finish, I used the _in_qtt and _out_qtt here
```sql
    IF @date_exists = 0 THEN
      SET _row_uuid  = HUID(uuid());
      INSERT INTO  `stock_movement_status` VALUES (_row_uuid, _start_date, _start_date, _qtt, IFNULL(_in_qtt, 0), IFNULL(_out_qtt, 0), _inventory_uuid, @depot_uuid);
    END IF;
```

mbayopanda commented on server/models/migrations/next/migrate.sql:
```diff
@@ -707,3 +707,26 @@ CREATE TABLE `lot_tag` (
+    CONSTRAINT `stock_movement_status__depot` FOREIGN KEY (`depot_uuid`) REFERENCES `depot` (`uuid`),
+    KEY `inventory_uuid` (`inventory_uuid`),
+    CONSTRAINT `stock_movment_status__inventory` FOREIGN KEY (`inventory_uuid`) REFERENCES `inventory` (`uuid`)
+) ENGINE=InnoDB DEFAULT CHARACTER SET = utf8mb4 DEFAULT COLLATE = utf8mb4_unicode_ci;
```
👍

mbayopanda reviewed (commented):
The stock entry and stock exit still working fine

We are writing successfully in the stock_movement_status

There is just one thing i noticed is `in_quantity` and `out_quantity` are not yet computed in the stock_movement_status table 

![image](https://user-images.githubusercontent.com/5445251/91861101-57ac0f00-ec64-11ea-8c91-dba522a21208.png)

jniles reviewed (commented):
bors r=mbayopanda,jmcameron,lomamech

jniles reviewed (commented):
bors r=jmcameron,mbayopanda,lomamech

jniles committed:
Merge remote-tracking branch 'upstream/master' into 4864-feat-compute-quanity-in-stock-proceedure
